### ISSUE TITLE
about precise loss about multihead fusion

### ISSUE THREAD

**Issue by lw921014** [HazyResearch/flash-attention#26]
Hi, Tri Tao,
    I test our [fwd kernel](https://github.com/HazyResearch/flash-attention/blob/main/csrc/flash_attn/fmha_api.cpp#L174)， I found we have about 2~3% precise loss. Do you find such loss?

**Comment by tridao** [comment]
We have [tests](https://github.com/HazyResearch/flash-attention/blob/2ed471ecc450801cbe30be2169936a7fe69058c4/tests/test_flash_attn.py#L520) that compare the numerical error of FlashAttention and a reference implementation, and the numerical error is similar to the Pytorch implementation.
These numerical errors are because we're using fpt16 / bf16.

Can you say more about what "2-3% precise loss" mean? Are you comparing the outputs of FlashAttention vs Pytorch implementation? Or are you comparing the end models (e.g. GPT2), one using FlashAttention and one using standard attention?

**Comment by lw921014** [comment]
> Can you say more about what "2-3% precise loss" mean? Are you comparing the outputs of FlashAttention vs Pytorch implementation?

Thanks for your reply. 

We do three comparison about precise, they are

1. a single kernel: we test fwd kernel. I mean this [kernel](https://github.com/HazyResearch/flash-attention/blob/main/csrc/flash_attn/fmha_api.cpp#L174)
2. a single layer with that in pytorch, I construct a swin-transformer window attention layer use pytorch and our flash atteniton, then compare.
3. a model in SWIN-T. I replace all the attention layer use impl in 2, then run the model. Note, although we ignore mask and pos bias in flash attention, so we ignore it in origin swin-t window attention layer. 

for all of three case, we all have 2-3% precise loss. I dont know why?

**Comment by lw921014** [comment]
> 2\. a single layer with that in pytorch, I construct a swin-transformer window attention layer use pytorch and our flash atteniton, then compare.
> 3\. a model in SWIN-T. I replace all the attention layer use impl in 2, then run the model. Note, although we ignore mask and pos bias in flash attention, so we ignore it in origin swin-t window attention layer.

we use apex O2 opt level, so all the data are fp16.

**Comment by tridao** [comment]
Using fp16 necessarily means there's numerical error for any operation (often in the range of 1e-3 for each operation).
One thing you can compare is the error:
|FlashAttention in fp16 - Reference attention in fp32| vs |Reference attention in fp16 - Reference attention in fp32|.
We test [here](https://github.com/HazyResearch/flash-attention/blob/2ed471ecc450801cbe30be2169936a7fe69058c4/tests/test_flash_attn.py#L520) that the two errors are around the same (for a single attention layer).

**Comment by lw921014** [comment]
> We have [tests](https://github.com/HazyResearch/flash-attention/blob/2ed471ecc450801cbe30be2169936a7fe69058c4/tests/test_flash_attn.py#L520) that compare the numerical error of FlashAttention and a reference implementation, and the numerical error is similar to the Pytorch implementation.

Do we test [this interface](https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attention.py#L80) using apex?

**Comment by tridao** [comment]
No, we don't have that test yet. You can check to see how the numerical error changes with 3 operations (matmul, attention, matmul).

**Comment by lw921014** [comment]
> No, we don't have that test yet. You can check to see how the numerical error changes with 3 operations (matmul, attention, matmul).

ok, Thanks a lot.

**Comment by lw921014** [comment]
> No, we don't have that test yet. You can check to see how the numerical error changes with 3 operations (matmul, attention, matmul).

I have do this test with model. You can see https://github.com/HazyResearch/flash-attention/pull/29, and if possible, can you help me check this process and whether some error exists? 

This test is about a single layer in swin-T, when I replace windows attention in swin-T using our flash attention, the final precise loss of all model is about 2-3%, which is very similar that for a single layer. I just want to know whether this case can be solved and got the expected precise of this swin-T.
